zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+/// is ephemeral at all, regardless of reuse.
+fn ephemeral_address_check_internal<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    address_str: &str,
+    reject_all_ephemeral: bool,
+) -> Result<(), SqliteClientError> {
+    // It is intentional that we don't require `t.mined_height` to be non-null.
+    // That is, we conservatively treat an ephemeral address as potentially
+    // reused even if we think that the transaction where we had evidence of
+    // its use is at present unmined. This should never occur in supported
+    // situations where only a single correctly operating wallet instance is
+    // using a given seed, because such a wallet will not reuse an address that
+    // it ever reserved.
+    //
+    // `COALESCE(used_in_tx, mined_in_tx)` can only differ from `used_in_tx`
+    // if the address was reserved, an error occurred in transaction creation
+    // before calling `mark_ephemeral_address_as_used`, and then we observed
+    // the address to have been used in a mined transaction (presumably by
+    // another wallet instance, or due to a bug) anyway.
+    let res = wdb
+        .conn
+        .0
+        .query_row(
+            "SELECT t.txid FROM ephemeral_addresses
+             LEFT OUTER JOIN transactions t
+             ON t.id_tx = COALESCE(used_in_tx, mined_in_tx)
+             WHERE address = :address",
+            named_params![":address": address_str],
+            |row| row.get::<_, Option<Vec<u8>>>(0),
+        )
+        .optional()?;
+
+    match res {
+        Some(Some(txid_bytes)) => {
+            let txid = TxId::from_bytes(
+                txid_bytes
+                    .try_into()
+                    .map_err(|_| SqliteClientError::CorruptedData("invalid txid".to_owned()))?,
+            );
+            Err(SqliteClientError::EphemeralAddressReuse(
+                address_str.to_owned(),
+                Some(txid),
+            ))
+        }
+        Some(None) if reject_all_ephemeral => Err(SqliteClientError::EphemeralAddressReuse(
+            address_str.to_owned(),
+            None,
+        )),
+        _ => Ok(()),
+    }
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in a transaction that we have just created. This has no effect if `address` is
+/// not one of our ephemeral addresses.
+///
+/// Returns a `SqliteClientError::EphemeralAddressReuse` error if the address was
+/// already used.
+pub(crate) fn mark_ephemeral_address_as_used<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    ephemeral_address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let address_str = encode_transparent_address_p(&wdb.params, ephemeral_address);
+    ephemeral_address_check_internal(wdb, &address_str, false)?;
+
+    wdb.conn.0.execute(
+        "UPDATE ephemeral_addresses SET used_in_tx = :used_in_tx WHERE address = :address",
```

Comment:
By comparison, there's no bug here, because `mark_ephemeral_address_as_used` should only be called on addresses that we've previously reserved (assuming no programming bugs elsewhere).